cloudinit/sources/DataSourceAzure.py

```diff
@@ -1275,24 +1293,36 @@ def _negotiate(self):
             self.bounce_network_with_azure_hostname()
 
             pubkey_info = None
+            keys_from_imds = True
             try:
-                raise KeyError(
-                    "Not using public SSH keys from IMDS"
-                )
-                # pylint:disable=unreachable
                 public_keys = self.metadata['imds']['compute']['publicKeys']
-                LOG.debug(
-                    'Successfully retrieved %s key(s) from IMDS',
-                    len(public_keys)
-                    if public_keys is not None
-                    else 0
-                )
+                for key in [key['keyData'] for key in public_keys]:
+                    if not _key_is_openssh_formatted(key=key):
+                        keys_from_imds = False
+                        break
+
+                if not keys_from_imds:
+                    log_msg = 'Keys not in OpenSSH format, using OVF'
+                else:
+                    log_msg = 'Retrieved {} keys from IMDS'.format(
+                        len(public_keys)
+                        if public_keys is not None
+                        else 0
+                    )
             except KeyError:
-                LOG.debug(
-                    'Unable to retrieve SSH keys from IMDS during '
-                    'negotiation, falling back to OVF'
-                )
+                log_msg = 'Unable to get keys from IMDS, falling back to OVF'
+                keys_from_imds = False
+            finally:
+                report_diagnostic_event(log_msg, logger_func=LOG.debug)
+
+            if not keys_from_imds:
                 pubkey_info = self.cfg.get('_pubkeys', None)
+                log_msg = 'Retrieved {} keys from OVF'.format(
+                    len(pubkey_info)
+                    if pubkey_info is not None
+                    else 0
+                )
+                report_diagnostic_event(log_msg, logger_func=LOG.debug)
```

Comment:
This logic all looks quite repetitious; is there something we could do to reduce the duplication with `get_public_ssh_keys`?

Reply:
Done! And good point. I refactored some of this logic out for deduplication.